### ISSUE TITLE
[v2 QUESTION]: Is cross validation and num folds currently not complete in chemprop v2?

### ISSUE THREAD

**Issue by davdma** [chemprop/chemprop#974]
**What are you trying to do?**
I would like to run cross validation with a random or scaffold split in my hyperparameter tuning and subsequent testing for benchmarking chemprop (as done in the paper). While there is a cross validation flag and num folds flag, I looked extensively at the source code, and it seems that even when cross validation is specified, the hyperparameter tuning script just takes the first split for each trial instead of running the specified num-folds number of splits and computing the average performance over those folds. Moreover, I also found in the data splitting function that no cross validation is done for SCAFFOLD_BALANCED split at all regardless of num-folds parameter. 

I just want to confirm if what I suspect is the case, and when the features will be implemented by Chemprop v2 as I do need these features for my benchmarking experiments. Unless maybe I am misunderstanding the source code.

**Comment by shihchengli** [comment]
Hi @davdma , your understanding is correct. I have also thought about this for a while, as in v1 we do follow this approach. cc @hwpang, as you have done the implementation of the hyperparameters search CLI, and cc @JacksonBurns for the `split_data()` function. I would be interested in your thoughts on this. Would it make sense for us to support these functionalities?

**Comment by davdma** [comment]
Hi @shihchengli, I wonder what reason you have for not supporting this functionality? I have a small dataset, so tuning just on one split can lead to very noisy results. I tested it out myself, and the performance of the tuned model parameters does not generalize well to other splits. In the meantime if this functionality is not going to be implemented, is there a simple way for me to implement it myself (@JacksonBurns @hwpang)?

**Comment by shihchengli** [comment]
The only thing I can think about is data leakage, as almost every data point would be seen during the HPO with multiple folds. We can rely on users to ensure there is a held-out test set that is not used during the HPO. I would be happy to put this functionality back, but I would also like to hear the other devs' opinions on this.

**Comment by davdma** [comment]
@shihchengli Hm that is a good point. But yes, maybe chemprop could then provide an option to hold out a test set and throughout the hpopt script just seed the train and validation sets across the number of folds. In the meantime, do you have recommendations for how I could use chemprop v2 for a small dataset (~400 molecules) as it currently stands for my situation and get consistent benchmarking results?

**Comment by shihchengli** [comment]
Ideally, I would suggest nested cross-validation to prevent data leakage, but it is not doable in our CLI and would require some effort to develop. For now, you can manually split your dataset into multiple splits and perform hyperparameter optimization for each split (you can provide the split via the `--splits-file` flag). The difference between nested cross-validation and this approach is that there is only one split in the inner loop. Moreover, try using ensemble models in model training via `--ensemble-size` to reduce model variance from your predictions.

**Comment by JacksonBurns** [comment]
I would actually encourage you to focus on multiple _replicates_ rather than cross validation.

Especially when training on small datasets, the dominant source of error (i.e. inconsistency between different models) is actually the contents of the test set. The benefit of 'striping' the validation set through the training data is much less than 'striping' the test set - beyond that, just doing multiple randomized test sets allows you to get the same result as actually rigorously ensuring each sample ends up in test at some point.

@shihchengli this is part of a greater discussion we need to have offline about how we should split data. My impression of the current literature is that cross-validation is losing out to replicates, and we should consider moving that way too. This is why the underlying splitting library `astartes` doesn't implement CV. Let's talk about this at the next subgroup.
